### DISCUSSION TITLE
Flattening Relational Databases to CSV

### DISCUSSION THREAD

**Discussion by rebmizrahi** [DDMAL/linkedmusic-datalake#5] [General]
After working on a draft of a script to convert simssadb to a CSV [(linked here)](https://github.com/rebmizrahi/linkedmusic-datalake/blob/main/SQL_query.py), we have a better idea of the general process + main challenges we will encounter flattening any databases. 
-  The SQL query requires explicitly stating each column to include (unless we use * which is probably unnecessary). Simssa has around around 30 tables to potentially be joined with multiple columns in each. 
-  We need to choose which columns to exclude. Other databases may also have more complicated table structures. Further, simssa has things like source, source instantiations, musical works, and files that are related, which are terms that need to be explained to understand. If other databases have nomenclature unique to them, extra information will be needed for narrowing things down. However, simssa seems to have a particularly complex data structure, so this may not be an issue at all.
- The query joins each table one by one with foreign key constraints. Constraints of a table can be listed with a simple psql command. 
- From what I understand, since we're only extracting triples/quads, no reorganizing of the table is necessary, which simplifies things.

Overall, the process is tedious but very doable, especially since the number of databases is small.

**Comment by malajvan** [comment]
The generated file is [here](https://github.com/rebmizrahi/linkedmusic-datalake/blob/main/output.tsv). Note that since simssadb values sometimes contain a list of values, it was not ideal to store in CSV format. Therefore we decided to generate a TSV instead.

**Reply by ahankinson** [reply]
CSV can accept commas in values as long as the string is quoted. Most libraries will do this automatically.

**Reply by malajvan** [reply]
Thank you very much, I checked and this was indeed automatically handled by Postgres and OpenRefine

**Comment by malajvan** [comment]
I did some more flattening and expansions (most notably flattening and pivoting the extracted features into columns). You can find the file and the python code in my fork: [malajvan/linkedmusic-datalake](https://github.com/malajvan/linkedmusic-datalake/tree/main/simssadb_flatten)
